### PR TITLE
Add DispatchingValueParser

### PR DESCRIPTION
**Pull request by thiemowmde**
This is split from and required for https://gerrit.wikimedia.org/r/#/c/197093/.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,6 +49,10 @@ DataValues Common has been written by the Wikidata team, as [Wikimedia Germany]
 
 ## Release notes
 
+### 1.0.0 (alpha)
+
+* Added `CompositeValueParser`
+
```
Comment (JeroenDeDauw) [inline README.md]: two alpha sections?

File: src/ValueParsers/CompositeValueParser.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace ValueParsers;
+
+use InvalidArgumentException;
+
+/**
+ * A generic value parser that forwards parsing to a list of other value parsers and returns the
+ * result of the first parse attempt that succeeded.
+ *
+ * @since 0.5
```
Comment (JeroenDeDauw) [inline src/ValueParsers/CompositeValueParser.php]: nope.jpg

File: src/ValueParsers/CompositeValueParser.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	 *
+	 * @throws ParseException
+	 * @return mixed
+	 */
+	public function parse( $value ) {
+		foreach ( $this->parsers as $parser ) {
+			try {
+				return $parser->parse( $value );
+			} catch ( ParseException $ex ) {
+				continue;
```
Comment (JeroenDeDauw) [inline src/ValueParsers/CompositeValueParser.php]: This is a rather icky. Generally you do not want to use exceptions for expected execution paths, which is happening here.

**Comment by JeroenDeDauw** [comment]
For what exactly is this composite parser being introduced?

**Comment by thiemowmde** [comment]
This is existing code, moved from `Wikibase\Lib\Parsers\TimeParser`.

Suggestions on how to implement this?

What about an other name for the class to better reflect it's behavior?

**Comment by JeroenDeDauw** [comment]
You can do what is done here: https://github.com/wmde/Serialization/tree/master/src/Serializers

**Comment by thiemowmde** [comment]
Isn't this rather pointless because there is no easy way to detect if a parser can parse a certain string, other than trying and checking the result somehow? Such an `isParserFor` method can't do much but call `parse` and either `catch` exceptions.

**Comment by JeroenDeDauw** [comment]
Indeed. You could apply a regrex. (Keep in mind the goal would be to find the correct parser, not to ensure the value is actually parse-able by it.) Not sure this makes sense. I need to check the existing usage first.

**Comment by thiemowmde** [comment]
Checking if it's worth to pass a string to a slightly more complex parser (as all TimeValue parsers are) is not possible with a regex. And even if, such a regex can't catch all cases and the try-catch can't be removed, so what's the benefit?

**Comment by JeroenDeDauw** [comment]
Checked the TimeParser now. I'm fine with this being added here, and using the try catch.

I understand correctly that the intent here is to create instances of the CompositeValueParser with other parsers and not to derive from it?

How about `DisplatchingValueParser`?

**Comment by thiemowmde** [comment]
Derive? We can make this a `final` class if that's what worries you. I just want to use it, as I already do in the linked https://gerrit.wikimedia.org/r/#/c/197093/.

Not sure about "dispatching". What I don't want is something that sounds like what you initially suggested, a delegating parser that first decides which parser to use and then delegates to this and only this parser. While writing this I realize "composition" is misleading. It doesn't "combine" parsers, it picks one and only one. I will rename.

**Comment by thiemowmde** [comment]
Renamed.

**Comment by JeroenDeDauw** [comment]
Ok, that answers my question yes. Making it final is good then yes.

**Comment by thiemowmde** [comment]
I meant it *could* be final but I'm not going to make it final. I don't plan to extend this class, but why should that be forbidden?

**Comment by JeroenDeDauw** [comment]
It makes intent more clear. I can't find fault with the reasoning here https://ocramius.github.io/blog/when-to-declare-classes-final/

Which does not mean I'm intent on making a lot of our things final. In this case there is particular ambiguity about intent though, so it'd be nice to clarify.

**Comment by thiemowmde** [comment]
"Particular ambiguity about intent"? I don't think so. It's a parser that dispatches parsing to other parsers. What's unclear about that? Sure, it's not obvious how exactly it decides which parser to use. But I don't see how this is a reason to block users of this component from extending the class. There are only 9 final classes across our code base. We have similar "dispatcher" classes. None of them is final.

**Comment by JeroenDeDauw** [comment]
> There are only 9 final classes across our code base. We have similar "dispatcher" classes. None of them is final.

You can see how that is a logical fallacy right?

**Comment by thiemowmde** [comment]
I find the "if it implements an interface, and no other public method" rule very nice, but: Consistency is not a fallacy. We can have the same discussion for quite a lot of code styling details:
* It seems to be a good idea to make everything `protected` by default. There are good arguments to do this, and others to not do it.
* In languages that support `final`, it's seems to be a good idea to make all method parameters `final` so they can not be reused for other purposes in the method. But even if that's "the right thing to do"<sup>TM</sup> it's ugly.
* It seems to be a good idea to make methods `static` if they do not access properties anyway, but there are arguments to not do this.

And so on.

**Comment by JeroenDeDauw** [comment]
Do you think that consistently doing something that has been shown to be non-optimal makes sense in a situation where there are no additional reasons to do that?

**Pull request merged**